lib/Dialect/Moore/Transforms/DeleteLocalVar.cpp

```diff
@@ -0,0 +1,123 @@
+//===- DeleteLocalVar.cpp - Delete local temporary variables --------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the DeleteLocalVar pass.
+// Use to collect net/variable declarations and bound a value to them.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetail.h"
+#include "circt/Dialect/Moore/MoorePasses.h"
+
+using namespace circt;
+using namespace moore;
+
+namespace {
+struct DeleteLocalVarPass : public DeleteLocalVarBase<DeleteLocalVarPass> {
+  void runOnOperation() override;
+};
+} // namespace
+
+namespace {
+struct DLV { // Delete Local Variables
+  /// All local temporary variables at the `moore.procedure`, `scf.if`,
+  /// `scf.while` body region/block need to be deleted, they won't be reserved
+  /// at the low level.
+  void deleteLocalVar(DenseSet<Operation *> &localVars,
+                      DenseSet<Operation *> &users, mlir::Region &region);
+
+  /// The `valueSymbols` used to collect nets/variables and their initial or
+  /// assignment values.
+  using ValueSymbols = llvm::ScopedHashTable<Operation *, Value>;
+  using ValueSymbolScope = ValueSymbols::ScopeTy;
+  ValueSymbols valueSymbols;
+};
+} // namespace
+
+void DLV::deleteLocalVar(DenseSet<Operation *> &localVars,
+                         DenseSet<Operation *> &users, mlir::Region &region) {
+  ValueSymbolScope scope(valueSymbols);
+
+  for (auto &op : region.getOps()) {
+    // Assume not to assign a new value to the local variable.
+    bool isNewValue = false;
+
+    TypeSwitch<Operation *, void>(&op)
+        // Collect all local variables and their initial values
+        // if exist and all users.
+        .Case<VariableOp>([&](auto op) {
+          valueSymbols.insert(op, op.getInitial());
+          localVars.insert(op);
+          for (auto *user : op->getUsers())
+            users.insert(user);
+        })
+        // Update the values of local variables.
+        .Case<BlockingAssignOp>([&](auto op) {
```

Comment:
nit: Does local variable are known/verified not to have non-blocking assignments?

Reply:
Ah I meant to ensure that there is no NonBlockingAssignOp for local variables since it wouldn't be correct